### ISSUE TITLE
Better way

### ISSUE THREAD

**Issue by momala454** [lawfx/YoutubeNonStop#25]
Hello, i analysed the youtube code and what is triggering the popup is on this file
https://www.youtube.com/yts/jsbin/desktop_polymer_inlined_html_polymer_flags_v2-vflNwouvq/desktop_polymer_inlined_html_polymer_flags_v2.js

There are multiple ways the popup/pause is triggered,
first there is the checking of the variable window._lact which is the timestamp of the last action. Modifying it regularily will prevent some of the trigger, but there are triggers that are called via ajax.

On the polymer file, there is this function
```
function(k){if(1==k.nextAddress){a.reset();a.youThereData_=b.configData.youThereData;if(a.youThereData_.reason&&"YOU_THERE_TRIGGER_REASON_3"==a.youThereData_.reason&&!jU())return EP("YOU_THERE_EVENT_TYPE_RATE_LIMITED"),k.return();c=parseInt(a.youThereData_.lactThresholdMs,10);return!B("disable_youthere_lact_threshold_check")&&Sn()+1E3*a.youThereData_.promptDelaySec<c?(EP("YOU_THERE_EVENT_TYPE_BELOW_LACT_THRESHOLD",Ko()),k.return()):0>=a.youThereData_.playbackPauseDelayMs?(ZCa(a),k.return()):xa(k,
a.playerProvider_().getPlayerPromise(),2)}d=k.yieldResult;e=d.getCurrentTime();if(!a.youThereData_)return k.return();g=a.youThereData_.promptDelaySec;a.showWarningJob_=Gn(L,function(){return Ka(function(m){a.showWarningJob_=0;return xa(m,$Ca(a),0)})},1E3*Math.max(g-e,0));
ya(k)}
```

i think that if you replace it with
`function(k){}`

it will not be triggered.
However, i don't know how to modify on the fly the script, i tried with tampermonkey to intercept the script and put the script content from the external file to an inline script, but it doesn't work, the code is not executed. Maybe you know how to do that ?

**Comment by momala454** [comment]
actually i found a way that seems to works after multiple hours of test.
Simply do

window.addEventListener('load', function() {
    document.querySelector('ytd-watch-flexy').youthereDataChanged_ = function() {};
}, false);

i put that into tampermonkey

**Comment by momala454** [comment]
i published a greasemonkey/tampermonkey script here https://greasyfork.org/fr/scripts/401348-yes-i-m-here-youtube

**Comment by lawfx** [comment]
Nice, I think you are on to something there. From a quick test I did, it indeed does seem to work, if you put the code directly into the console but I tried it as an extension and from there it seems to not recognize these variables youthereDataChanged_, youThereManager_. I would have to look more into it.

**Comment by momala454** [comment]
follow my greasyfork link, i have published an updated version, but now it's my own code haha :)

**Comment by lawfx** [comment]
Yeah I saw it, that's the code I tried. Don't worry, I'm not gonna use your code in any case :D 
I don't really want to override any YouTube functions anyway, since I can't be sure what code they execute themselves and I don't want to break anything in the process.

**Comment by momala454** [comment]
on which browser did you tried

**Comment by lawfx** [comment]
Chrome

**Comment by momala454** [comment]
did you tried with tampermonkey + install the script ?

**Comment by Storm-Engineer** [comment]
> i published a greasemonkey/tampermonkey script here https://greasyfork.org/fr/scripts/401348-yes-i-m-here-youtube

Thank you for your script! This addon is broken for a long time and the owner ignores my reports but your script works flawlessly so I just uninstalled the addon and use the script. Thanks again!

**Comment by Josehkn** [comment]
> i published a greasemonkey/tampermonkey script here https://greasyfork.org/fr/scripts/401348-yes-i-m-here-youtube

Hey!!! Thanks worked for me

**Comment by BPower0036** [comment]
@momala454 The solution proposed by @Araxeus can easily implemented in every extension, as I have done.
Otherwise you need two extensions tampermonkey and in my case an autoskip extension.
 (https://github.com/BPower0036/YT-Nonstop)

**Comment by Mlocik97** [comment]
About that pausing, I was testing it without any extensions. Normally I have only AdGuard and GNOME in Chrome, and last few months I have this extension. This extension works fine, except month ago there was problem, it was impossible to pause video  (see: https://github.com/lawfx/YoutubeNonStop/issues/41 ) so I was trying to make my own extension based on combination of few solutions. I also tried using momala's script, but it didn't worked (video paused anyway), Not sure if it's difference between YouTube and YTMusic, I tested it only on YTMusic. Then I tried another methods with few scripts I wrote and just dumbly pasted to console. Some worked halfway fine, some didn't. But then issue with pausing was solved in this extension so I again started using this extension. Btw. I tested reseting `window.__lact` and it works fine too.

**Comment by momala454** [comment]
> 
> 
> > use my tampermonkey script it works better :)
> 
> how exactly does it work better? why is yours even doing so much stuff every second?
> 
> I really tested my way (_lact=timeNow every 15min) for hundreds of hours and it worked perfectly

what do you mean so much stuff every second ? The first timeout is only to wait that the player is loaded, then it remove a function and never execute later

**Comment by Mlocik97** [comment]
it doesn't make sense to use setInterval that way... why not use setTimeout ?

**Comment by momala454** [comment]
No I'm only running it 100 times if the player have not been found. I've never had any problem on YouTube, but I didn't tested on YouTube music if that was your problem

**Comment by Mlocik97** [comment]
> No I'm only running it 100 times if the player have not been found.

and isn't it bad anyway...  and performance is not worse...

> your solution have "worse performance" as you run it every 15 minutes while I run it only once or a few times until the player have been loaded and it never run after

30 or 100 times but every second, not every 15 minutes... it's worse performance... because these 100 iterations are done in 1 minute and 40 seconds, in which it makes amount of instructions, that solution with `windows.__lact` makes only after hours... so no, You solution doesn't have better performance, mainly in first minute and 40 seconds or maybe resp. in 30 seconds...

**Comment by momala454** [comment]
You forgot what I said about popup starting directly when loading a video. Modifying that variable after 15 minutes will not works. I wanted something that works for me in every case

**Comment by Mlocik97** [comment]
Your solution didn't worked as well. At least on YT Music.

**Comment by momala454** [comment]
I never tested on music, the popup on start happened multiple times for me, and it happened a few times that the player wasn't loaded so I had to add a retry functionality but using timeout or interval doesn't make a big difference as I'm clearing the interval once the player is loaded. It's just a preference, but that doesn't make the code bad. I'm running every second as the popup that display on start was only a few seconds after loading so it was required for me.

**Comment by JohnyP36** [comment]
@Araxeus  The code I proposed in https://github.com/lawfx/YoutubeNonStop/issues/25#issuecomment-954668455 and that you proposed in https://github.com/lawfx/YoutubeNonStop/pull/43#issuecomment-954702962 (which is Manifest v3) makes for some reason the extension stop working.  
How do I know? 
Because after youtube is loaded the extension just vanished in my browser task management (Shift+Esc). 

I don't know if this could be the cause the extension not working.

**Comment by BPower0036** [comment]
@JohnyP36 This is https://github.com/BPower0036/YT-Nonstop still working

**Comment by JohnyP36** [comment]
How can you add @momala454 's script into lawfx 's extension? 

> actually i found a way that seems to works after multiple hours of test. Simply do
> 
> window.addEventListener('load', function() { document.querySelector('ytd-watch-flexy').youthereDataChanged_ = function() {}; }, false);
> 
> i put that into tampermonkey

Do you have to change the `autoconfirm.js` to this: 
```
const script = document.createElement('script');
script.textContent = window.addEventListener('load', function() { document.querySelector('ytd-watch-flexy').youthereDataChanged_ = function() {}; }, false);
script.onload = function () {
    this.remove();
};
(document.head || document.documentElement).appendChild(script);
```
or maybe this: 
```
var ynsInjection =
  '(' +
  function () {
    const tag = '[YT Nonstop]';
    const isYoutubeMusic = window.location.hostname === 'music.youtube.com';
    function log(message) {
      console.log(`${tag}[${getTimestamp()}] ${message}`);
    }
    function getTimestamp() {
      let dt = new Date();
      let time = dt.getHours() + ':' + dt.getMinutes() + ':' + dt.getSeconds();
      return time;
    }
    window.addEventListener('load', function() { document.querySelector('ytd-watch-flexy').youthereDataChanged_ = function() {}; }, false);
  } +
  ')();';
console.log(`[Youtube NonStop v${chrome.runtime.getManifest().version}]`);
var script = document.createElement('script');
script.textContent = ynsInjection;
(document.head || document.documentElement).appendChild(script);
script.remove();
```

**Comment by jimmywarting** [comment]
at least i can say that i do not like the mutation observer... 
I rather much prefer a interval that changes last modified action timestamp on a global property

listening on all DOM changes are more costly i think

**Comment by Mlocik97** [comment]
I must disagree with jimmy.

yes, listening to DOM changes is costly, but we don't need to listen to whole DOM. And actually if we listen to just specific part of DOM that doesn't change often, it's for sure better than interval, that would either execute too often unnecesary, or miss some change, as time of change is not known. Also cost of listener is actually not that big, as cost of whole logic, that is triggered by it. Ofc, for making it work efficiently, we shouldn't use mutation observer, but just inject listener on only that part of DOM, that has effect on video auto-pause.  So if specific elements internal attributes change, just rewrite them at that moment.
